### PR TITLE
Added Static CustomOp with lowering to regular custom Op

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**
Currently, the classical parameters of quantum gates are compiled dynamically even if they are known at compile time. We would like the IR to be extended to support literal values as opposed to SSA Values for such static parameters.

**Description of the Change:**
This is achieved by adding a new type of gate called StaticCustomOp which is similar to CustomOp except that it uses a DenseF64ArrayAttr which lists all the literal values for parameters in square bracket. For example the following IR: 
```
%c = llvm.mlir.constant(2.000000e-01 : f64)
%result = quantum.custom "RX"(%c) %qubit
```
would change to:
```
%result = quantum.static_custom "RX" [2.000000e-01] %qubit
```
The static custom ops are then lowered to the regular custom ops after apply-transform-sequence pass for the rest of the compilation process.
Currently we have **not** supported Multirz, and GlobalShift.

**Benefits:**
More flexibility in case there is a lack of support for dynamic quantum circuits.

**Possible Drawbacks:**

**Related GitHub Issues:**
Static parameters [sc-73581]

### CODE REVIEW

**Comment by mehrdad2m** [comment]
> This looks a lot better! You'll also need to change the merge rotations and cancel inverses passes right?

Thanks @erick-xanadu, yes I am planning to change those today.

File: mlir/lib/Quantum/IR/QuantumOps.cpp
Comment (mehrdad2m) [reply mlir/lib/Quantum/IR/QuantumOps.cpp:46]: Yes I agree with this. I think I just copied over customOp canonicalize method here. Didn't look too much into it :) I will fix this as well

File: frontend/catalyst/jax_extras/tracing.py
Comment (dime10) [inline frontend/catalyst/jax_extras/tracing.py:992]: I wonder if we just allow lists in general here? Would avoid including this special case with the `static_params` name. 
Also some minor simplification of the code (no need for dictionary update syntax with only 1 element, can remove early exit since branch is still legal & inexpensive if list is empty):
```suggestion
            static_arg[flex_arg_name] = flex_arg_value
        elif isinstance(flex_arg_value, list):
            if all(type(arg) in static_literal_pool for arg in flex_arg_value):
                static_args[flex_arg_name] = flex_arg_value
```

Comment (dime10) [reply frontend/catalyst/jax_extras/tracing.py:992]: Either let's update the docstring of this function to reflect the changes :)

**Review by dime10 — CHANGES_REQUESTED** [review]
Nice work! I know we also discussed the MultiRZ/GlobalPhase handling in the frontend so I'll wait for those for final approval, but otherwise nothing major 👍

File: mlir/lib/Quantum/IR/QuantumOps.cpp
Comment (dime10) [reply mlir/lib/Quantum/IR/QuantumOps.cpp:46]: > > This looks a lot better! You'll also need to change the merge rotations and cancel inverses passes right?
> Thanks @erick-xanadu, yes I am planning to change those today.

I also don't think these need to be updated to run on the static ops since they can run after the conversion to dynamic.

Comment (mehrdad2m) [reply mlir/lib/Quantum/IR/QuantumOps.cpp]: With the current design, the lowering is done after the apply-transform-sequence pass which is ultimately the place where peephole optimizations are run, so passes like cancel inverses and merge rotations should be able to deal with static custom ops. I think the reason for this placement in pipeline is to allow external contributors to schedule their passes within the apply-transform-sequence pass.

File: frontend/catalyst/from_plxpr.py
Comment (mehrdad2m) [reply frontend/catalyst/from_plxpr.py:308]: the reason for this change is that after using using bind_flexible_primitive, the gate parameters will be appended to the list of all dynamic parameters and hence it will change the order of arguments. previously it was [qubits, params, ctrl_qubits, ctrl_values], but after the change it is [qubits, ctrl_qubits, ctrl_values, params]. so we need to know the length of ctrl_values instead of params.

File: frontend/catalyst/pipelines.py
Comment (mehrdad2m) [reply frontend/catalyst/pipelines.py:173]: As far as I know, we expect changes from mqt to happen in apply-transform-sequence pass. That's why it is placed right after that.

File: mlir/include/Quantum/IR/QuantumOps.td
Comment (mehrdad2m) [reply mlir/include/Quantum/IR/QuantumOps.td:483]: Yes, the brackets would be printed anyways. I didn't spend time to make a custom printer, but will consider it if I have time before the end of week

File: mlir/lib/Quantum/Transforms/StaticCustomPatterns.cpp
Comment (mehrdad2m) [reply mlir/lib/Quantum/Transforms/StaticCustomPatterns.cpp:37]: Oh Actually `op.getStaticParams() ` returns a `llvm::ArrayRef<double>` instead of the attribute which is convenient for looping over :) to get the attribute one should use op.getStaticParamsAttr();

File: mlir/include/Quantum/IR/QuantumOps.td
Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td:483]: Don't worry I think it's alright 👍

File: mlir/lib/Quantum/Transforms/StaticCustomPatterns.cpp
Comment (dime10) [reply mlir/lib/Quantum/Transforms/StaticCustomPatterns.cpp:37]: Oh neat, thanks :)

File: frontend/catalyst/pipelines.py
Comment (dime10) [reply frontend/catalyst/pipelines.py:173]: Oh good point, that's where the plugins are applied. All good then 👍

File: mlir/lib/Quantum/IR/QuantumOps.cpp
Comment (dime10) [reply mlir/lib/Quantum/IR/QuantumOps.cpp]: Ah I see, oh well then I guess it's fine 👍

File: frontend/catalyst/from_plxpr.py
Comment (dime10) [reply frontend/catalyst/from_plxpr.py:308]: Oh I see, the plxr converter doesn't support dynamic control qubits yet, so the 0 makes sense.

I just realized though that `ctrl_len` and `ctrl_value_len` should always be the same, so we might not need this variable and/or assert that it they are the same.

**Comment by mehrdad2m** [comment]
> Nice work! I know we also discussed the MultiRZ/GlobalPhase handling in the frontend so I'll wait for those for final approval, but otherwise nothing major 👍

Thanks @dime10 for the comments. I added two lit tests for the lowering pass and addressed all of the comments except for "MultiRZ/GlobalPhase handling in the frontend". I will work on it next, but maybe it is safer to do that in another PR? just to make sure this one get merged by the end of the week. what do you think?

File: frontend/catalyst/jax_extras/tracing.py
Comment (mehrdad2m) [reply frontend/catalyst/jax_extras/tracing.py]: I applied the suggestions except for the early exit. if I don't have the early exit, I would get a {'static_param' : []} in the static arguments which causes some tests that compare JAXPRs to fail.

File: frontend/catalyst/from_plxpr.py
Comment (dime10) [reply frontend/catalyst/from_plxpr.py:308]: Still feel weird about the `ctrl_value_len` parameter being here, semantically it just feels out of place / not that meaningful. Can we just get rid of it?

File: mlir/lib/Quantum/Transforms/static_custom_lowering.cpp
Comment (dime10) [inline mlir/lib/Quantum/Transforms/static_custom_lowering.cpp:49]: Don't think this is necessary for a partial conversion, since it will only fail if explicitly illegal ops cannot be converted.
```suggestion
```

**Review by dime10 — APPROVED** [review]
Nice work 💯 Happy for the odd gates to come in a follow up 👍

File: frontend/catalyst/from_plxpr.py
Comment (mehrdad2m) [reply frontend/catalyst/from_plxpr.py:308]: oh I just remembered that I added this to pass a test. There are some tests in `frontend/test/pytest/test_from_plxpr.py` that compare the jaxpr from plpxr and catalyst. If I remove the `ctrl_value_len` then that test fails because one will have the `ctrl_value_len` in the jaxpr and the other will not. But otherwise I agree that there is not point of having it there.

File: mlir/lib/Quantum/Transforms/static_custom_lowering.cpp
Comment (mehrdad2m) [reply mlir/lib/Quantum/Transforms/static_custom_lowering.cpp]: target.markUnknownOpDynamicallyLegal was needed since the newly added `custom` and `arith.constant` operations that replace the `static_ustom` operation would fail to legalize. I removed this and legalized those two operations explicitly to avoid confusion.

Comment (dime10) [reply mlir/lib/Quantum/Transforms/static_custom_lowering.cpp]: Thanks, that's helpful!

**Pull request merged**
